### PR TITLE
fix: double parsing issue in sas9 when debug is enabled

### LINKED ISSUE #506: double parsing parseWeboutResponse
we need to remove these line from `WebJobExecutor.ts` file.
Screenshot from 2021-08-10 18-54-14

this scenario is being handled from parseResponse method of RequestClient class.

### PR DESCRIPTION
**Pull request by sabhas**
## Issue

closes #506

## Intent

What this PR intends to achieve.
-  refactor code regarding getValidJson function
- check valid json while parsing sas viya debug response
- resolve parsing issue in sas9 when debug is enabled

## Implementation

- introduced new error types
- refactored getValidJson function with proper error types.
- also updated tests for getValidJson function
- called getValidJson function before returing res.result from parseSASViyaDebugResponse
- in webJobExecutor when serverType is sas9 and debug is true check if res.result is string then call parseWeboutResponse function

## Checks

No PR (that involves a non-trivial code change) should be merged, unless all items below are confirmed!  If an urgent fix is needed - use a tar file.

- [x] All `sasjs-cli` unit tests are passing (`npm test`).
Mocked:
image
Server:
image
- [x] All `sasjs-tests` are passing (instructions available [here](https://github.com/sasjs/adapter/blob/master/sasjs-tests/README.md)).
image
- [x] [Data Controller](https://datacontroller.io) builds and is functional on both SAS 9 and Viya
SAS Viya:
image
SAS9:
image

### CODE REVIEW

File: src/job-execution/WebJobExecutor.ts [hunk 1]
```diff
@@ -102,29 +103,30 @@ export class WebJobExecutor extends BaseJobExecutor {
...
           if (this.serverType === ServerType.Sas9 && config.debug) {
-            const jsonResponse = parseWeboutResponse(res.result as string)
-            if (jsonResponse === '') {
-              throw new Error(
-                'Valid JSON could not be extracted from response.'
-              )
+            let jsonResponse
+            if (typeof res.result === 'string') {
+              jsonResponse = parseWeboutResponse(res.result)
+              if (jsonResponse === '') throw new WeboutResponseError(apiUrl)
```
Comment (YuryShkoda) [inline src/job-execution/WebJobExecutor.ts:120]: I think that it makes more sense to throw error from `parseWeboutResponse`

File: src/types/errors/JsonParseArrayError.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export class JsonParseArrayError extends Error {
+  constructor(public message: string) {
+    super(message)
```
Comment (YuryShkoda) [inline src/types/errors/JsonParseArrayError.ts:3]: In opinion, error message should be predefined

File: src/utils/getValidJson.ts [hunk 3]
```diff
@@ -1,16 +1,19 @@
...
-    }
+    const arrayErrorMessage = 'Can not parse array object to json.'
+    if (Array.isArray(str)) throw new JsonParseArrayError(arrayErrorMessage)
+
     if (typeof str === 'object') return str
 
     return JSON.parse(str)
   } catch (e) {
-    throw new Error('Invalid JSON response.')
+    if (e instanceof JsonParseArrayError) throw e
```
Comment (YuryShkoda) [inline src/utils/getValidJson.ts:16]: We' ve agreed to style such statements differently

**Review by YuryShkoda — CHANGES_REQUESTED** [review]
see comments above

File: src/job-execution/WebJobExecutor.ts (hunk 1 above)
Comment (sabhas) [reply src/job-execution/WebJobExecutor.ts:120]: we wouldn't have access to apiUrl in parseWeboutResponse function.

**Review by YuryShkoda — APPROVED** [review]

**Review by VladislavParhomchik — APPROVED** [review]

**Pull request merged**